velox/exec/Window.cpp

```diff
@@ -57,21 +57,68 @@ void Window::initialize() {
   windowNode_.reset();
 }
 
+namespace {
+void checkRowFrameBounds(const core::WindowNode::Frame& frame) {
+  auto frameBoundCheck = [&](const core::TypedExprPtr& frameValue) -> void {
+    if (frameValue == nullptr) {
+      return;
+    }
+
+    VELOX_USER_CHECK(
+        frameValue->type() == INTEGER() || frameValue->type() == BIGINT(),
+        "k frame bound must be INTEGER or BIGINT type");
+  };
+  frameBoundCheck(frame.startValue);
+  frameBoundCheck(frame.endValue);
+}
+
+void checkKRangeFrameBounds(
+    const std::shared_ptr<const core::WindowNode>& windowNode,
+    const core::WindowNode::Frame& frame,
+    const RowTypePtr& inputType) {
+  // For k Range frame bound:
+  // i) The order by needs to be a single column for bound comparisons
+  // (Checked in WindowNode constructor).
+  // ii) The bounds values are pre-computed in the start(end)Value bound
+  // fields. So, start(end)Value bounds cannot be constants.
+  // iii) The frame bound column and the ORDER BY column must have
+  // the same type for correct comparisons.
+  auto orderByType = windowNode->sortingKeys()[0]->type();
+  auto frameBoundCheck = [&](const core::TypedExprPtr& frameValue) -> void {
+    if (frameValue == nullptr) {
+      return;
+    }
+
+    auto frameChannel = exprToChannel(frameValue.get(), inputType);
+    VELOX_USER_CHECK_NE(
+        frameChannel,
+        kConstantChannel,
+        "Window frame of type RANGE does not support constant arguments");
```

Comment:
Hi @aditi-pandit, I'm assuming `frame.startValue/endValue` references to the columns 
produced by Presto to store lower/upper bound values for each row. But for some frameworks like Spark, no such columns is produced. So it needs Velox to compute the upper/lower bound value for each row based on the passed constant `startValue`/`endValues` (e.g., 4 or 2 for `range between 4 preceding and 2 following`).

We can extend Velox to support no-precompute frameworks based on your impl. in the future, right? Then, I guess we need to change the check logic here to allow constant argument, and introduce some extra handling for such cases, e.g., compute the lower/upper bound value for each row.

Reply:
@PHILO-HE : It is definitely possible to support a variation without pre-compute. 

There are many cases to handle for it however (integer, date, intervals etc). It would be great if you can provide more details about the options needed for Spark. Based on that we could design further.

Reply:
@aditi-pandit, yes, computing lower/upper bound for each row needs to handle different cases. Just checked spark source. When sort column is Date/Timestamp type, the constant `startValue`/`endValue` can be `IntervalDayTimeType`/`IntervalYearMonthType`, which makes the computing different from most primitive types. And timezone should also be considered in the computing. But I think it is OK to firstly support commonly used basic types.

Reply:
@PHILO-HE : What are the basic types you have in mind ?

Reply:
I think integer types, float & double should be the mostly used types.

Reply:
@PHILO-HE : Had another question for you : Does Spark support columns for start/end k range frame bounds e.g. range between c1 preceding and current row ?

Reply:
> Does Spark support columns for start/end k range frame bounds e.g. range between c1 preceding and current row?

No, Spark doesn't support columns for start/end range frame bound. They must be constant. So maybe, we can infer pre-computing is already done if column is used. Otherwise, we need to let velox compute the start/end bound. 

Just created an issue for discussion:
https://github.com/facebookincubator/velox/issues/7557

Reply:
@PHILO-HE It might be cleaner to use a projection before window in the query plan passed to Velox to precompute these columns.

Reply:
Resolved in https://github.com/facebookincubator/velox/issues/7557